### PR TITLE
optimization: make `RedundantJoin` work on non-column expressions

### LINKED ISSUE #9049: optimization: detect redundant joins on non-column expressions
```
materialize=> explain select * from t1, (select distinct f1 % 2 as f from t1) t where t1.f1 % 2 = t.f;
                Optimized Plan                
----------------------------------------------
 %0 = Let l0 =                               +
 | Get materialize.public.t1 (u7)            +
 | Filter !(isnull(#0))                      +
                                             +
 %1 =                                        +
 | Get %0 (l0)                               +
 | Project (#0)                              +
 | Distinct group=((#0 % 2))                 +
 | ArrangeBy (#0)                            +
                                             +
 %2 =                                        +
 | Join %0 %1 (= #2 (#0 % 2))                +
 | | implementation = Differential %0 %1.(#0)+
 
(1 row)
```

The query above could simplified as:

```
materialize=> explain select * from t1, (select distinct f1 % 2 as f from t1) t where t1.f1 % 2 = t.f;
                Optimized Plan                
----------------------------------------------
 %0 =                                 +
 | Get materialize.public.t1 (u7)            +
 | Filter !(isnull(#0))                      +
 | Map (#0 % 2)
 
(1 row)
```

However, the current redundant join transform only detect redundancies based on column expressions.

### PR DESCRIPTION
**Pull request by asenac**
### Motivation

Fixes #9049.

This PR extends `RedundantJoin` transform to work on arbitrary expressions. Instead of lifting the source column each column in the projection of the current operator corresponds to, it lifts the underlying expression behind each projected column written in terms of columns of the source of `Get` operator of the `ProvInfo` instance.

With that information we can check whether two elements in a join equivalence lead to the same underlying source expression.

### Tips for reviewer

Comments are addressed in `fixup` commits that will be eventually squashed once the discussion around them settles.

### Checklist

- [x] This PR has adequate test coverage / QA involvement has been duly considered.
- [ ] This PR adds a release note for any
  [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/user/content/release-notes.md#what-changes-require-a-release-note).

### CODE REVIEW

**Comment by philip-stoev** [comment]
Item No #1. This query is optimized so well that the entire error-producing expression and the join clause is thrown out:

```
materialize=> EXPLAIN SELECT ( SELECT MIN(f1) FROM t2) FROM (SELECT DISTINCT f1 / 0 AS f1 FROM t1) AS outer1 , t1 AS outer2 WHERE outer2 . f1 / 0 = outer1 . f1;

 %0 = Let l0 =                             +
 | Get materialize.public.t2 (u89)         +
 | Project (#0)                            +
 | Reduce group=()                         +
 | | agg min(#0)                           +
                                           +
 %1 =                                      +
 | Get %0 (l0)                             +
 | Project ()                              +
 | Negate                                  +
                                           +
 %2 =                                      +
 | Constant ()                             +
                                           +
 %3 =                                      +
 | Union %1 %2                             +
 | Map null                                +
                                           +
 %4 = Let l1 =                             +
 | Union %0 %3                             +
                                           +
 %5 =                                      +
 | Get materialize.public.t1 (u85)         +
 | Filter !(isnull(#0))                    +
 | Project ()                              +
 | ArrangeBy ()                            +
                                           +
 %6 =                                      +
 | Get %4 (l1)                             +
 | Project ()                              +
 | Negate                                  +
                                           +
 %7 =                                      +
 | Constant ()                             +
                                           +
 %8 =                                      +
 | Union %6 %7                             +
 | Map null                                +
                                           +
 %9 =                                      +
 | Union %4 %8                             +
                                           +
 %10 =                                     +
 | Join %5 %9                              +
 | | implementation = Differential %9 %5.()+
```

There is no `/0` Filter anywhere in the query and the join produced on step %10 does not have a `/0`  join condition anymore.

I assume this is by design. I guess it comes back to the age-old question about how much to optimize things away, which I do not mean to resurrect.

**Review by philip-stoev — APPROVED** [review]
I am able to see this optimization in action -- there were no wrong results and panics. The query plans that I examined were all net positive.

Thank you for the tests!

File: src/transform/src/redundant_join.rs [hunk 1]
```diff
@@ -390,50 +479,73 @@ fn find_redundancy(
...
+                                    try_build_expression_using_other(
+                                        root_expr,
+                                        other,
+                                        other_prov,
+                                        input_mapper,
+                                    )
+                                } else {
+                                    None
+                                }
                             })
```
Comment (wangandi) [inline src/transform/src/redundant_join.rs:544]: The `if ... let...` can be replaced with `root_expr.map_or(...)`

File: src/transform/src/redundant_join.rs [hunk 2]
```diff
@@ -341,28 +348,110 @@ impl RedundantJoin {
...
+                .map(|c| Some(MirScalarExpr::column(c)))
+                .collect::<Vec<_>>(),
+            exact: true,
+        }
+    }
+
+    /// Given an expression on top of the projection of the current operator
+    /// it returns an equivalent expression as if it was right on top of the
+    /// underlying Get operator.
+    fn dereference(&self, expr: &MirScalarExpr) -> Option<MirScalarExpr> {
```
Comment (wangandi) [inline src/transform/src/redundant_join.rs:374]: Maybe we can add `try_visit_mut` (see the corresponding method in src/expr/src/relation/mod.rs) and use `try_visit_mut` in this method and in `try_build_expression_using_other`. Whenever, this method would have returned None, throw an error. But then again, using `try_visit_mut` would require cloning `expr`.

File: src/transform/src/redundant_join.rs [hunk 3]
```diff
@@ -341,28 +348,110 @@ impl RedundantJoin {
 pub struct ProvInfo {
```
[the remark this answers is not in the document]
Comment (asenac) [reply src/transform/src/redundant_join.rs:348]: Ooops, thanks. I missed that.

File: test/sqllogictest/transform/redundant_join.slt [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+# Use of this software is governed by the Business Source License
+# included in the LICENSE file at the root of this repository.
+#
+# As of the Change Date specified in that file, in accordance with
+# the Business Source License, use of this software will be governed
+# by the Apache License, Version 2.0.
+
+statement ok
+CREATE TABLE t1(f1 INT, f2 INT);
+----
```
[the remark this answers is not in the document]
Comment (asenac) [reply test/sqllogictest/transform/redundant_join.slt:12]: TIL! Thanks! I might have committed other tests with the dashed lines. I'll check.

File: src/transform/src/redundant_join.rs [hunk 5]
```diff
@@ -390,50 +479,73 @@ fn find_redundancy(
...
+                                // as the key column.
                                 equivalences.iter().any(|e| {
-                                    e.contains(&input_mapper.map_expr_to_global(
-                                        MirScalarExpr::Column(*input_col),
-                                        input,
-                                    )) && e.contains(&input_mapper.map_expr_to_global(
-                                        MirScalarExpr::Column(other_col),
-                                        other,
-                                    ))
+                                    e.iter().any(|e| {
```
[the remark this answers is not in the document]
Comment (asenac) [reply src/transform/src/redundant_join.rs]: :joy: yeah, way too many `e`s in that line :)

File: src/transform/src/redundant_join.rs [hunk 6]
```diff
@@ -112,7 +109,7 @@ impl RedundantJoin {
                 // We only do this once per invocation to keep our sanity, but we could
                 // rewrite it to iterate. We can avoid looking for any relation that
                 // does not have keys, as it cannot be redundant in that case.
-                if let Some((input, bindings)) = (0..input_types.len())
+                if let Some((input, mut bindings)) = (0..input_types.len())
```
[the remark this answers is not in the document]
Comment (asenac) [reply src/transform/src/redundant_join.rs]: yeah, definitely clearer

File: src/transform/src/redundant_join.rs [hunk 7]
```diff
@@ -127,51 +124,65 @@ impl RedundantJoin {
...
+                                        *c -= old_input_mapper.input_arity(remove_input_idx);
+                                    }
+                                }
+                            });
                         }
                     }
+
                     expr::canonicalize::canonicalize_equivalences(equivalences, &input_types);
 
+                    let new_input_mapper = JoinInputMapper::new_from_input_types(&input_types);
```
Comment (wangandi) [inline src/transform/src/redundant_join.rs:165]: Add a comment about what this block does.

**Review by wangandi — APPROVED** [review]

**Pull request merged**